From 136d1a367a2d6fa29aca8b44818ceb070af59c61 Mon Sep 17 00:00:00 2001
From: Venkateswara Naralasetty <vnaralas@codeaurora.org>
Date: Wed, 7 Apr 2021 14:38:54 +0530
Subject: [PATCH] sigma-dut-open: set mu beamformer support based on hw capab

Currently mu beamformer support is set by default for HE program
from ap_reset_default(). Due to this AP DUT fails to brinup for
the hardwares which are not MU beamformer capable in AP mode
like IPQ5018.

Signed-off-by: Venkateswara Naralasetty <vnaralas@codeaurora.org>
---
 ap.c           | 44 ++++++++++++++++++++++++++++++++++++++++++++
 nl80211_copy.h | 17 +++++++++++++++++
 sigma_dut.h    |  2 ++
 3 files changed, 63 insertions(+)

--- a/ap.c
+++ b/ap.c
@@ -8222,11 +8222,13 @@ static int fill_dual_radio_caps(struct s
 		if (dut->ap_sgi80 || dut->ap_txBF ||
 		    dut->ap_ldpc != VALUE_NOT_SET ||
 		    dut->ap_tx_stbc || dut->ap_mu_txBF ||
-		    dut->ap_ampdu_exp || dut->ap_chwidth_1) {
-			len += snprintf(vht_capab + len, sizeof(vht_capab) - len, "%s%s%s%s%s%s",
+		    dut->ap_ampdu_exp || dut->ap_chwidth_1 ||
+		    dut->ap_su_txBF) {
+			len += snprintf(vht_capab + len, sizeof(vht_capab) - len, "%s%s%s%s%s%s%s",
 					dut->ap_sgi80 ? "[SHORT-GI-80]" : "",
 					dut->ap_txBF ?
-					"[SU-BEAMFORMER][SU-BEAMFORMEE][BF-ANTENNA-2][SOUNDING-DIMENSION-2]" : "",
+					"[SU-BEAMFORMER][BF-ANTENNA-2][SOUNDING-DIMENSION-2]" : "",
+					dut->ap_su_txBF ? "[SU-BEAMFORMEE]":"",
 					(dut->ap_ldpc == VALUE_ENABLED) ?
 					"[RXLDPC]" : "",
 					dut->ap_tx_stbc ? "[TX-STBC-2BY1][RX-STBC-1]" : "",
@@ -9398,11 +9400,12 @@ skip_key_mgmt:
 		    dut->ap_ldpc != VALUE_NOT_SET ||
 		    dut->ap_tx_stbc || dut->ap_mu_txBF ||
 		    dut->ap_chwidth || dut->ap_ampdu_exp ||
-		    dut->ap_max_mpdu_len) {
-			len += snprintf(vht_capab + len, sizeof(vht_capab) - len, "%s%s%s%s%s%s",
+		    dut->ap_max_mpdu_len || dut->ap_su_txBF) {
+			len += snprintf(vht_capab + len, sizeof(vht_capab) - len, "%s%s%s%s%s%s%s",
 				dut->ap_sgi80 ? "[SHORT-GI-80]" : "",
 				dut->ap_txBF ?
-				"[SU-BEAMFORMER][SU-BEAMFORMEE][BF-ANTENNA-2][SOUNDING-DIMENSION-2]" : "",
+				"[SU-BEAMFORMER][BF-ANTENNA-2][SOUNDING-DIMENSION-2]" : "",
+				dut->ap_su_txBF ? "[SU-BEAMFORMEE]":"",
 				(dut->ap_ldpc == VALUE_ENABLED) ?
 				"[RXLDPC]" : "",
 				dut->ap_tx_stbc ? "[TX-STBC-2BY1][RX-STBC-1]" : "",
@@ -10247,9 +10250,42 @@ static void phy_info_vht_capa(struct dut
 	}
 }
 
+static void phy_info_he_capa(struct dut_hw_modes *mode,
+			     struct nlattr *nl_iftype)
+{
+	struct nlattr *tb[NL80211_BAND_IFTYPE_ATTR_MAX + 1];
+	struct nlattr *tb_flags[NL80211_IFTYPE_MAX + 1];
+	int len;
+
+	nla_parse(tb, NL80211_BAND_IFTYPE_ATTR_MAX,
+		  nla_data(nl_iftype), nla_len(nl_iftype), NULL);
+
+	if (!tb[NL80211_BAND_IFTYPE_ATTR_IFTYPES])
+		return;
+
+	if (nla_parse_nested(tb_flags, NL80211_IFTYPE_MAX,
+			     tb[NL80211_BAND_IFTYPE_ATTR_IFTYPES], NULL))
+		return;
+
+	if (!nla_get_flag(tb_flags[NL80211_IFTYPE_AP]))
+		return;
+
+	if (tb[NL80211_BAND_IFTYPE_ATTR_HE_CAP_PHY]) {
+		len = nla_len(tb[NL80211_BAND_IFTYPE_ATTR_HE_CAP_PHY]);
+
+		if (len > ARRAY_SIZE(mode->ap_phy_capab) - 1)
+			len = ARRAY_SIZE(mode->ap_phy_capab) - 1;
+
+		memcpy(mode->ap_phy_capab,
+		       nla_data(tb[NL80211_BAND_IFTYPE_ATTR_HE_CAP_PHY]), len);
+	}
+}
+
 static int phy_info_band(struct dut_hw_modes *mode, struct nlattr *nl_band)
 {
 	struct nlattr *tb_band[NL80211_BAND_ATTR_MAX + 1];
+	struct nlattr *nl_iftype;
+	int rem_band;
 
 	if (!mode)
 		return 1;
@@ -10265,6 +10301,13 @@ static int phy_info_band(struct dut_hw_m
 	phy_info_vht_capa(mode, tb_band[NL80211_BAND_ATTR_VHT_CAPA],
 			  tb_band[NL80211_BAND_ATTR_VHT_MCS_SET]);
 
+	if (tb_band[NL80211_BAND_ATTR_IFTYPE_DATA]) {
+		nla_for_each_nested(nl_iftype,
+				    tb_band[NL80211_BAND_ATTR_IFTYPE_DATA],
+				    rem_band)
+		phy_info_he_capa(mode, nl_iftype);
+	}
+
 	/* Other nl80211 band attributes can be parsed here, if required */
 
 	return NL_OK;
@@ -10299,15 +10342,58 @@ static int wiphy_info_handler(struct nl_
 			return res;
 	}
 
+	if (dut->program == PROGRAM_HE && dut->hw_modes &&
+	    !(dut->hw_modes->ap_phy_capab[4] & HE_PHY_CAP_MU_BEAMFORMER))
+		dut->ap_mu_txBF = 0;
+
+	if (dut->program == PROGRAM_HE && dut->hw_modes &&
+	    !(dut->hw_modes->vht_capab & IEEE80211_VHT_CAP_SU_BEAMFORMEE_CAPABLE))
+		dut->ap_su_txBF = 0;
+
 	return 0;
 }
 
+static int protocol_feature_handler(struct nl_msg *msg, void *arg)
+{
+	u32 *feat = arg;
+	struct nlattr *tb_msg[NL80211_ATTR_MAX + 1];
+	struct genlmsghdr *gnlh = nlmsg_data(nlmsg_hdr(msg));
+
+	nla_parse(tb_msg, NL80211_ATTR_MAX, genlmsg_attrdata(gnlh, 0),
+		  genlmsg_attrlen(gnlh, 0), NULL);
+
+	if (tb_msg[NL80211_ATTR_PROTOCOL_FEATURES])
+		*feat = nla_get_u32(tb_msg[NL80211_ATTR_PROTOCOL_FEATURES]);
+
+	return NL_SKIP;
+}
+
+static u32 get_nl80211_protocol_features(struct sigma_dut *dut, int ifindex)
+{
+	u32 feat = 0;
+	struct nl_msg *msg;
+
+	if(!(msg = nl80211_drv_msg(dut, dut->nl_ctx, ifindex, 0,
+				   NL80211_CMD_GET_PROTOCOL_FEATURES))) {
+		sigma_dut_print(dut, DUT_MSG_DEBUG,
+				"%s: could not get protocol feature", __func__);
+		return 0;
+	}
+
+	if (send_and_recv_msgs(dut, dut->nl_ctx, msg, protocol_feature_handler,
+			       &feat) == 0)
+		return feat;
+
+	return 0;
+}
 
 static int mac80211_get_wiphy(struct sigma_dut *dut)
 {
 	struct nl_msg *msg;
 	int ret = 0;
 	int ifindex;
+	u32 feat;
+	int flags = 0;
 
 	ifindex = if_nametoindex(dut->main_ifname);
 	if (ifindex == 0) {
@@ -10317,9 +10403,14 @@ static int mac80211_get_wiphy(struct sig
 		return -1;
 	}
 
-	if (!(msg = nl80211_drv_msg(dut, dut->nl_ctx, ifindex, 0,
+	feat = get_nl80211_protocol_features(dut, ifindex);
+	if (feat & NL80211_PROTOCOL_FEATURE_SPLIT_WIPHY_DUMP)
+		flags = NLM_F_DUMP;
+
+	if (!(msg = nl80211_drv_msg(dut, dut->nl_ctx, ifindex, flags,
 				    NL80211_CMD_GET_WIPHY)) ||
-	    nla_put_u32(msg, NL80211_ATTR_IFINDEX, ifindex)) {
+	    nla_put_u32(msg, NL80211_ATTR_IFINDEX, ifindex) ||
+	    nla_put_flag(msg, NL80211_ATTR_SPLIT_WIPHY_DUMP)) {
 		sigma_dut_print(dut, DUT_MSG_DEBUG,
 				"%s: could not build get wiphy cmd", __func__);
 		nlmsg_free(msg);
@@ -10394,6 +10485,7 @@ static enum sigma_cmd_result cmd_ap_rese
 	dut->ap_rx_amsdu = VALUE_NOT_SET;
 	dut->ap_txBF = 0;
 	dut->ap_mu_txBF = 0;
+	dut->ap_su_txBF = 1;
 	dut->ap_chwidth = AP_AUTO;
 	dut->ap_ampdu_exp = 7;
 	dut->ap_max_mpdu_len = 0;
--- a/nl80211_copy.h
+++ b/nl80211_copy.h
@@ -2942,6 +2942,22 @@ enum nl80211_mpath_info {
 	NL80211_MPATH_INFO_MAX = __NL80211_MPATH_INFO_AFTER_LAST - 1
 };
 
+enum nl80211_band_iftype_attr {
+       __NL80211_BAND_IFTYPE_ATTR_INVALID,
+
+       NL80211_BAND_IFTYPE_ATTR_IFTYPES,
+       NL80211_BAND_IFTYPE_ATTR_HE_CAP_MAC,
+       NL80211_BAND_IFTYPE_ATTR_HE_CAP_PHY,
+       NL80211_BAND_IFTYPE_ATTR_HE_CAP_MCS_SET,
+       NL80211_BAND_IFTYPE_ATTR_HE_CAP_PPE,
+       NL80211_BAND_IFTYPE_ATTR_HE_6GHZ_CAPA,
+
+       /* keep last */
+       __NL80211_BAND_IFTYPE_ATTR_AFTER_LAST,
+       NL80211_BAND_IFTYPE_ATTR_MAX = __NL80211_BAND_IFTYPE_ATTR_AFTER_LAST - 1
+};
+
+
 /**
  * enum nl80211_band_attr - band attributes
  * @__NL80211_BAND_ATTR_INVALID: attribute number 0 is reserved
@@ -2972,6 +2988,7 @@ enum nl80211_band_attr {
 
 	NL80211_BAND_ATTR_VHT_MCS_SET,
 	NL80211_BAND_ATTR_VHT_CAPA,
+	NL80211_BAND_ATTR_IFTYPE_DATA,
 
 	/* keep last */
 	__NL80211_BAND_ATTR_AFTER_LAST,
--- a/sigma_dut.h
+++ b/sigma_dut.h
@@ -125,6 +125,9 @@ struct wfa_p2p_attribute {
 #define VHT_CAP_MAX_A_MPDU_LENGTH_EXPONENT_MASK       \
                 (7 << VHT_CAP_MAX_A_MPDU_LENGTH_EXPONENT_SHIFT)
 
+#define IEEE80211_VHT_CAP_SU_BEAMFORMEE_CAPABLE 0x00001000
+
+#define HE_PHY_CAP_MU_BEAMFORMER 0x2
 
 #define AMPDU_FACTOR_BITS	0x3
 #define AMPDU_DENSITY_OFFSET	2
@@ -135,6 +138,7 @@ struct dut_hw_modes {
         u8 ampdu_params;
         u32 vht_capab;
         u8 vht_mcs_set[8];
+	u8 ap_phy_capab[11];
 };
 
 #define WPA_GET_BE32(a) ((((u32) (a)[0]) << 24) | (((u32) (a)[1]) << 16) | \
@@ -677,6 +681,7 @@ struct sigma_dut {
 	} ap_vht_chwidth;
 	int ap_txBF;
 	int ap_mu_txBF;
+	int ap_su_txBF;
 	enum ap_regulatory_mode {
 		AP_80211D_MODE_DISABLED,
 		AP_80211D_MODE_ENABLED,
